server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,218 @@
+            return history;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = history.isEmpty() ? null : history.get(history.size() - 1);
+        history = history.stream().filter(timeAndMaster -> timeAndMaster.time > thirtyMinutesAgo).collect(Collectors.toList());
+        if (history.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            history.add(mostRecent);
+        }
+        return history;
+    }
+
+    /**
+     * Clears out anything from masterHistory that is from more than 30 minutes before now (but leaves the newest entry in even if it is
+     * more than 30 minutes old). Rather than being scheduled, this method is called whenever the cluster state changes.
+     */
+    private void removeOldMasterHistory(List<TimeAndMaster> newMasterHistory) {
+        if (newMasterHistory.size() < 2) {
+            return;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = newMasterHistory.isEmpty() ? null : newMasterHistory.get(newMasterHistory.size() - 1);
+        newMasterHistory.removeIf(timeAndMaster -> timeAndMaster.time < thirtyMinutesAgo);
+        if (newMasterHistory.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            newMasterHistory.add(mostRecent);
+        }
+    }
+
+    /**
+     * This method returns an immutable view of this master history, typically for sending over the wire to another node.
+     * @return An immutable view of this master history
+     */
+    public List<DiscoveryNode> getImmutableView() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        return masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).toList();
+    }
+
+    private static class TimeAndMaster {
+        private final long time;
+        private final DiscoveryNode master;
+
+        TimeAndMaster(long time, DiscoveryNode master) {
+            this.time = time;
+            this.master = master;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) return true;
+            if (other instanceof TimeAndMaster == false) {
+                return false;
+            }
+            TimeAndMaster otherTimeAndMaster = (TimeAndMaster) other;
+            return time == otherTimeAndMaster.time
+                && ((master == null && otherTimeAndMaster.master == null) || (master != null && master.equals(otherTimeAndMaster.master)));
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(time, master);
+        }
+
+    }
```

Comment:
Hmm, way back at the beginning of this PR that's what it looked like. I had to change it for some reason, but I wouldn't be surprised if that reason is gone now. I'll re-evaluate.

Reply:
It might be because I was sending this over the wire as part of the transport action and had to make this implement ToXContent, or something related to that? I can't remember.